### PR TITLE
RavenDB-18701 - System.IO.IOException should be treated as out of dis…

### PR DESCRIPTION
**Pull request by shaharhikri**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18701/Error-should-be-treated-as-out-of-disk-space-error

### Additional description

System.IO.IOException should be treated as out of disk space exception.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Indexing/TempFileCache.cs [hunk 1]
```diff
@@ -214,8 +215,27 @@ public override void SetLength(long value)
...
+            try
+            {
+                InnerStream.Write(buffer, offset, count);
+                _length = Math.Max(_length, pos + count);
+            }
+            catch (IOException e)
+            {
+                if (IsDiskFull(e))
+                {
+                    throw new OutOfDiskSpaceException("Out of disk space", e);
```
Comment (grisha-kotler) [inline src/Raven.Server/Indexing/TempFileCache.cs]: You can use `DiskFullException` instead

File: src/Raven.Server/Indexing/TempFileCache.cs [hunk 2]
```diff
@@ -214,8 +215,27 @@ public override void SetLength(long value)
...
-            InnerStream.Write(buffer, offset, count);
-            _length = Math.Max(_length, pos + count);
+            try
+            {
+                InnerStream.Write(buffer, offset, count);
+                _length = Math.Max(_length, pos + count);
+            }
+            catch (IOException e)
+            {
+                if (IsDiskFull(e))
```
Comment (grisha-kotler) [inline src/Raven.Server/Indexing/TempFileCache.cs]: See `IsOutOfDiskSpaceException` in `ExceptionHelper`.

File: src/Raven.Server/Indexing/TempFileCache.cs [hunk 3]
```diff
@@ -214,8 +215,27 @@ public override void SetLength(long value)
...
+                {
+                    throw new OutOfDiskSpaceException("Out of disk space", e);
+                }
+            }
+        }
+
+        private static bool IsDiskFull(Exception ex)
+        {
+            const int HR_ERROR_HANDLE_DISK_FULL = unchecked((int)0x80070027);
```
Comment (grisha-kotler) [inline src/Raven.Server/Indexing/TempFileCache.cs]: This should be added in `IsOutOfDiskSpaceException` (for `Windows` only)

File: src/Raven.Server/Indexing/TempFileCache.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Indexing/TempFileCache.cs]: And we need to provide more info about the out-of-disk space error.
Like in `ThrowDiskFullException` in `VoronIndexOutput`

File: src/Raven.Server/Indexing/TempFileCache.cs [hunk 4]
```diff
@@ -214,8 +215,27 @@ public override void SetLength(long value)
         public override void Write(byte[] buffer, int offset, int count)
         {
             var pos = InnerStream.Position;
-            InnerStream.Write(buffer, offset, count);
-            _length = Math.Max(_length, pos + count);
+            try
+            {
+                InnerStream.Write(buffer, offset, count);
```
Comment (grisha-kotler) [inline src/Raven.Server/Indexing/TempFileCache.cs:228]: Need to apply the this also for `SetLength`

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Indexing/TempFileCache.cs [hunk 5]
```diff
@@ -214,8 +215,27 @@ public override void SetLength(long value)
...
+            catch (IOException e)
+            {
+                if (IsDiskFull(e))
+                {
+                    throw new OutOfDiskSpaceException("Out of disk space", e);
+                }
+            }
+        }
+
+        private static bool IsDiskFull(Exception ex)
```
Comment (ayende) [inline src/Raven.Server/Indexing/TempFileCache.cs]: What about Linux?

File: src/Raven.Server/Indexing/TempFileCache.cs (hunk 3 above)
Comment (shaharhikri) [reply src/Raven.Server/Indexing/TempFileCache.cs]: Added  ERROR_HANDLE_DISK_FULL.

File: src/Raven.Server/Indexing/TempFileCache.cs (hunk 5 above)
Comment (shaharhikri) [reply src/Raven.Server/Indexing/TempFileCache.cs]: The method IsOutOfDiskSpaceException does this already, so I used it instead.

File: src/Raven.Server/Indexing/TempFileCache.cs (hunk 1 above)
Comment (shaharhikri) [reply src/Raven.Server/Indexing/TempFileCache.cs]: Done

File: src/Raven.Server/Indexing/TempFileCache.cs (hunk 2 above)
Comment (shaharhikri) [reply src/Raven.Server/Indexing/TempFileCache.cs]: Done

File: src/Raven.Server/Indexing/TempFileCache.cs (hunk 4 above)
Comment (shaharhikri) [reply src/Raven.Server/Indexing/TempFileCache.cs:228]: Done

**Review by grisha-kotler — APPROVED** [review]

**Comment by shaharhikri** [comment]
fixed warnings

File: src/Raven.Server/Indexing/TempFileCache.cs [hunk 6]
```diff
@@ -1,12 +1,15 @@
...
 using System.Collections.Concurrent;
 using System.IO;
+using Raven.Server.Exceptions;
 using Raven.Server.ServerWide;
 using Raven.Server.Utils;
 using Sparrow.Global;
 using Sparrow.LowMemory;
 using Sparrow.Utils;
 using Voron;
+using Raven.Server.Utils;
```
Comment (karmeli87) [inline src/Raven.Server/Indexing/TempFileCache.cs]: please remove this

File: src/Raven.Server/Indexing/TempFileCache.cs [hunk 7]
```diff
@@ -205,17 +208,34 @@ public override long Seek(long offset, SeekOrigin origin)
...
-            _length = value;
+                _length = value;
+            }
+            catch (IOException e)
+            {
+                if (e.IsOutOfDiskSpaceException())
+                {
+                    IndexingUtils.ThrowDiskFullException(InnerStream.Name);
+                }
+            }
```
Comment (ppekrol) [inline src/Raven.Server/Indexing/TempFileCache.cs:220]: you are not rethrowing an exception when this is not out of disk space one, is this intended?

File: src/Raven.Server/Utils/IndexingUtils.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Sparrow.Server.Exceptions;
+using Sparrow.Server.Utils;
+
+namespace Raven.Server.Utils
+{
+    public static class IndexingUtils
+    {
+        public static void ThrowDiskFullException(string path) // Can be the folder path of the fole absolute path
```
Comment (ppekrol) [inline src/Raven.Server/Utils/IndexingUtils.cs]: better to move it to exception helper

File: src/Raven.Server/Indexing/TempFileCache.cs (hunk 6 above)
Comment (shaharhikri) [reply src/Raven.Server/Indexing/TempFileCache.cs]: Done

File: src/Raven.Server/Indexing/TempFileCache.cs (hunk 7 above)
Comment (shaharhikri) [reply src/Raven.Server/Indexing/TempFileCache.cs:220]: Done

File: src/Raven.Server/Utils/IndexingUtils.cs (hunk 8 above)
Comment (shaharhikri) [reply src/Raven.Server/Utils/IndexingUtils.cs]: Done

**Pull request merged**